### DISCUSSION TITLE
rendering smoke using add_volume

### DISCUSSION THREAD

**Discussion by chen112p** [pyvista/pyvista#3685] [Q&A] [answered]
Hi, I am trying to render smoke in a volume, opacity based on the concentration field value. I have made some success with Paraview, but i also want to know how to do it with pyvista. Right now, it seems add_volume() method only accept uniform grids. The time series data files are output from Openfoam fluid simulation so it is a unstructured grid. 

Is there a way to plot internal cells of the unstructured grid? I also tried to map concentration data to the uniform grid (20x20x60 nodes), but when using add_mesh crashed python kernel. I feel this is probably not the best way to do what i want to do...

**Comment by MatthewFlamm** [comment]
I'm interested in other people's thoughts on this for workarounds as well.  There is a definite feature request for better `add_volume` behavior.

I have used multiple isosurfaces with transparency to give a psuedo-volume rendered effect in the past.

**Reply by banesullivan** [reply]
>I'm interested in other people's thoughts on this for workarounds as well. There is a definite feature request for better add_volume behavior.

I've been wanting to improve `add_volume` for a while, but it would be a lot of work and would need some clear uses cases (like this) and someone to help fund this effort (something I'm looking into)

**Reply by banesullivan** [reply]
>I have used multiple isosurfaces with transparency to give a psuedo-volume rendered effect in the past.

This can actually be a great alternative for many datasets

**Reply by chen112p** [reply]
thanks for your input! thats a good idea. all i want to do here is just producing some eye candies and surprise ppl

**Reply by chen112p** [reply]
i also tried something similar..  taking multiple cut planes using concentration reading as opacity. i feel isosurface smoke is probably more real

**Comment by banesullivan** [comment]
Could you share these data to see what we can do with them? And to also reproduce what you are doing in ParaView?

**Reply by chen112p** [reply]
VTU file from openfoam output:
[smoke1.zip](https://github.com/pyvista/pyvista/files/10187137/smoke1.zip)

Using paraview, I can process this with the "Volume" option, and result looks like below. 
![image](https://user-images.githubusercontent.com/39268233/206495599-2d677c8c-7dae-4406-a395-d2163d4338e4.png)

**Reply by chen112p** [reply]
im actually also wondering if this can be done with a GPU

**Comment by banesullivan** [comment, marked as answer]
For posterity, volume rendering of unstructured grids was added in #3930
